File: Blind75/152. Maximum Product Subarray/152. Maximum Product Subarray.cpp
class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size();
        int minValue = 1;
        int maxValue = nums[0];
        int res = 0;
        for(int i=1; i<n; i++){
            if(nums[i]==0){
                if(i+1<n)minValue = nums[i+1];
                if(i+1<n) maxValue = nums[i+1];
                i++;
            }else if(nums[i]<0){
                minValue = minValue * nums[i];
            }else{
                maxValue = maxValue*nums[i];
            }

            res = max(res, max(minValue, maxValue));
        }
        
        return res;
    }
};

File: Blind75/152. Maximum Product Subarray/152. Maximum Product Subarray.cs
public class Solution {
    public int MaxProduct(int[] nums) {
        int n = nums.Length;
        int curr_max = nums[0];
        int curr_min = nums[0];
        int res = nums[0];

        for(int i=1; i<n; i++){
            int tmp = curr_min;
            curr_min = Math.Min(nums[i], Math.Min(nums[i]*curr_max, nums[i]*curr_min));
            curr_max = Math.Max(nums[i], Math.Max(nums[i]*curr_max, nums[i]*tmp));
            res = Math.Max(res, Math.Max(curr_min, curr_max));
        }

        return res;
    }
}
